File: 1590. Make Sum Divisible by P/solution.py
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        target_sum = sum(nums) % p
        n = len(nums)
        min_len = n
        curr_sum = 0
        
        if target_sum == 0:
            return 0

        remainder_idx = {0: -1}

        for i in range(n):
            curr_sum = (curr_sum + nums[i]) % p
            diff = (curr_sum - target_sum + p) % p

            if diff in remainder_idx:
                min_len = min(min_len, i - remainder_idx[diff])

            remainder_idx[curr_sum] = i

        return min_len if min_len != n else -1
        

File: 1590. Make Sum Divisible by P/solution.java
class Solution {
    public int minSubarray(int[] nums, int p) {

        long totalSum = 0;
        for (int num : nums) {
            totalSum += num;
        }

        long targetSum = totalSum % p;
        int n = nums.length, minLen = n;
        long currSum = 0;

        if (targetSum == 0) {
            return 0;
        }

        Map<Long, Integer> remainderIdx = new HashMap<>();
        remainderIdx.put(0L, -1);

        for (int i = 0; i < n; i++) {
            currSum = (currSum + nums[i]) % p;
            long diff = (currSum - targetSum + p) % p;

            if (remainderIdx.containsKey(diff)) {
                minLen = Math.min(minLen, i - remainderIdx.get(diff));
            }

            remainderIdx.put(currSum, i);
        }

        return minLen == n ? -1 : minLen;  
    }
}